### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by rockb1017**
RLP gateway logging and error channel is connected so that user can see any error logs from RLP connection.

### CODE REVIEW

File: eventsource/v2adapter.go [hunk 1]
```diff
@@ -29,7 +31,10 @@ func NewV2Adapter(s Streamer) V2Adapter {
 // Firehose returns only selected event stream
 func (a V2Adapter) Firehose(subscriptionID string) chan *events.Envelope {
 	ctx := context.Background()
-
+	var v1msgs = make(chan *events.Envelope, 1000)
```
Comment (luckyj5) [inline eventsource/v2adapter.go:34]: Why did we update this to 1000?

Comment (rockb1017) [reply eventsource/v2adapter.go:34]: just running experiment with how it impacts performance. let's have a meeting tmrw. let me share the results

File: eventsource/v2adapter.go [hunk 2]
```diff
@@ -61,17 +66,33 @@ func (a V2Adapter) Firehose(subscriptionID string) chan *events.Envelope {
...
 
-	var msgs = make(chan *events.Envelope, 100)
 	go func() {
 		for ctx.Err() == nil {
 			for _, e := range es() {
-				// ToV1 converts v2 envelopes down to v1 envelopes.
-				for _, v1e := range conversion.ToV1(e) {
-					msgs <- v1e
-				}
+				// consume events in channel and pull again
```
Comment (luckyj5) [inline eventsource/v2adapter.go:72]: Does this add any overhead? pulling again?
Are we planning to add both v1 and v2 event info as logging? or it's just for debugging?

File: splunknozzle/config.go [hunk 3]
```diff
@@ -139,7 +143,8 @@ func NewConfigFromCmdFlags(version, branch, commit, buildos string) *Config {
 		OverrideDefaultFromEnvar("ENABLE_EVENT_TRACING").Default("false").BoolVar(&c.TraceLogging)
 	kingpin.Flag("debug", "Enable debug mode: forward to standard out instead of splunk").
 		OverrideDefaultFromEnvar("DEBUG").Default("false").BoolVar(&c.Debug)
-
+	kingpin.Flag("mem-queue-monitor-interval", "Log the size of memory queue at every interval (in seconds)").
```
Comment (luckyj5) [inline splunknozzle/config.go:146]: in just seconds? or it accepts s/m/h? If yes, let's update the description.

File: splunknozzle/nozzle.go [hunk 4]
```diff
@@ -93,27 +97,29 @@ func (s *SplunkFirehoseNozzle) EventSink(logger lager.Logger) (eventsink.Sink, e
...
+		TraceLogging:            s.config.TraceLogging,
+		ExtraFields:             parsedExtraFields,
+		UUID:                    nozzleUUID,
+		Logger:                  logger,
+		MemqueueMonitorInterval: s.config.MemQueueMonitorInterval,
 	}
 
 	splunkSink := eventsink.NewSplunk(writers, sinkConfig)
-	if err := splunkSink.Open(); err != nil {
+	if err := splunkSink.Open(); err != nil { //this will never be hit
```
Comment (luckyj5) [inline splunknozzle/nozzle.go:122]: please remove/ update as needed.

File: splunknozzle/config.go (hunk 3 above)
Comment (rockb1017) [reply splunknozzle/config.go:146]: it accepts all time duration. will update description

File: splunknozzle/nozzle.go (hunk 4 above)
Comment (rockb1017) [reply splunknozzle/nozzle.go:122]: okay

File: testing/integration/test-env.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export SPLUNK_URL=https://ec2-3-15-168-6.us-east-2.compute.amazonaws.com:8088
```
Comment (luckyj5) [inline testing/integration/test-env.sh:1]: Please update this file and remove credentials.

File: eventsource/v2adapter.go [hunk 6]
```diff
@@ -27,11 +27,12 @@ func NewV2Adapter(s Streamer) V2Adapter {
 }
 
 // Firehose returns only selected event stream
-func (a V2Adapter) Firehose(subscriptionID string) chan *events.Envelope {
+func (a V2Adapter) Firehose(config *FirehoseConfig) chan *events.Envelope {
 	ctx := context.Background()
-
+	var v1msgs = make(chan *events.Envelope, 10000)
```
Comment (luckyj5) [inline eventsource/v2adapter.go:32]: Just curious, did we compare results between 100 and 10000 values?

Comment (luckyj5) [reply eventsource/v2adapter.go:32]: Do we need 2 separate channels here? Just wondering if additional channel will be sharing the same queue?

Comment (rockb1017) [reply eventsource/v2adapter.go:32]: to remove the duration it takes to convert it to v1. the way library is implemented, we have to call `es()` function as soon as possible. so i separated v2 msg channel and v1 msg channel.

Comment (rockb1017) [reply eventsource/v2adapter.go:32]: image

Comment (rockb1017) [reply eventsource/v2adapter.go:32]: Actually the size of this channel doesn't have huge difference. the `es` channel RLP library makes need to be changed. it is hard coded to 100 but it is too small. i definitely see unnecessary delays with 100 (screenshot above).

Comment (rockb1017) [reply eventsource/v2adapter.go:32]: I made another commit that change it to 10000. and these stay as 10000 just to match the size. the exact appropriate number should be reviewed once it is fixed on the PCF side. I think it is good for now.

File: splunknozzle/nozzle.go [hunk 7]
```diff
@@ -109,92 +115,107 @@ func (s *SplunkFirehoseNozzle) EventSink(logger lager.Logger) (eventsink.Sink, e
...
 		}
 		shutdownChan <- os.Interrupt
 	}()
+	select {
+	case <-shutdownChan:
+		s.logger.Info("Splunk Nozzle is going to exit gracefully")
+		noz.Close()
+		return eventSink.Close()
+	case gatewayError := <-gatewayErrChan:
+		s.logger.Error("Error from PCF RLP gateway", gatewayError)
```
Comment (luckyj5) [inline splunknozzle/nozzle.go:215]: Please update "PCF RLP" to "Reverse log proxy" and also add context to when do nozzle get this error.

File: splunknozzle/config.go [hunk 8]
```diff
@@ -128,6 +129,8 @@ func NewConfigFromCmdFlags(version, branch, commit, buildos string) *Config {
 		OverrideDefaultFromEnvar("CONSUMER_QUEUE_SIZE").Default("10000").IntVar(&c.QueueSize)
 	kingpin.Flag("hec-batch-size", "Batchsize of the events pushing to HEC").
 		OverrideDefaultFromEnvar("HEC_BATCH_SIZE").Default("100").IntVar(&c.BatchSize)
+	kingpin.Flag("rlp-gateway-retries", "Number of retries to connect to PCF RLP gateway").
```
Comment (luckyj5) [inline splunknozzle/config.go:132]: Update "PCF RLP" to Reverse log proxy

File: splunknozzle/config.go [hunk 9]
```diff
@@ -128,6 +129,8 @@ func NewConfigFromCmdFlags(version, branch, commit, buildos string) *Config {
...
+		OverrideDefaultFromEnvar("RLP_GATEWAY_RETRIES").Default("10").IntVar(&c.RLPGatewayRetries)
```
Comment (luckyj5) [inline splunknozzle/config.go:133]: Does this 10 mean number of retries? 10 seems high for default value. What is the time lag between different retries?

**Pull request merged**